alibi_detect/cd/pytorch/lsdd_online.py

```diff
@@ -0,0 +1,201 @@
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'. Only relevant for 'pytorch' backend.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            ert=ert,
+            window_size=window_size,
+            preprocess_x_ref=preprocess_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            n_bootstraps=n_bootstraps,
+            n_kernel_centers=n_kernel_centers,
+            lambda_rd_max=lambda_rd_max,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'pytorch'})
+
+        # set backend
+        if device is None or device.lower() in ['gpu', 'cuda']:
+            self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        self._configure_normalization()
+
+        # initialize kernel
+        if sigma is None:
+            self.x_ref = torch.from_numpy(self.x_ref).to(self.device)
+            self.kernel = GaussianRBF()
+            _ = self.kernel(self.x_ref, self.x_ref, infer_sigma=True)
+        else:
+            sigma = torch.from_numpy(sigma).to(self.device) if isinstance(sigma, np.ndarray) else None
+            self.kernel = GaussianRBF(sigma)
+
+        self._configure_kernel_centers()
+        self._configure_thresholds()
+        self._initialise()
+
+    def _configure_normalization(self):
+        x_ref_means = self.x_ref.mean(axis=0)
+        x_ref_stds = self.x_ref.std(axis=0)
+        self._normalize = lambda x: (x - x_ref_means)/(x_ref_stds)
+        self.x_ref = self._normalize(self.x_ref)
+
+    def _configure_kernel_centers(self):
+        "Set aside reference samples to act as kernel centers"
+        perm = torch.randperm(self.n)
+        self.c_inds, self.non_c_inds = perm[:self.n_kernel_centers], perm[self.n_kernel_centers:]
+        self.kernel_centers = self.x_ref[self.c_inds]
+        self.x_ref_eff = self.x_ref[self.non_c_inds]  # the effective reference set
+        self.k_xc = self.kernel(self.x_ref_eff, self.kernel_centers)
+
+    def _configure_thresholds(self):
+
+        # Each bootstrap sample splits the reference samples into a sub-reference sample (x)
+        # and an extended test window (y). The extended test window will be treated as W overlapping
+        # test windows of size W (so 2W-1 test samples in total)
+
+        w_size = self.window_size
+        etw_size = 2*w_size-1  # etw = extended test window
+        nkc_size = self.n - self.n_kernel_centers  # nkc = non-kernel-centers
+        rw_size = nkc_size - etw_size  # rw = ref-window
+        d = self.x_ref_eff.shape[-1]
```

Comment:
Not the only place where d is inferred like this, but it occurs that in the case where `self.x_ref` has not undergone preprocessing yet (`preprocess_x_ref=False` and `preprocess_fn` is not None) this could lead to issues if the shape of `x_ref` is not simply `(nb instances, nb features)` but eg `(nb instances, C, H, W)` for images. Might be best to raise an error in the online base class if `not preprocess_x_ref and isinstance(preprocess_fn, Callable)` or even remove `preprocess_x_ref` altogether in the first place.